spec/p2p/reactor/p2p-api.md

```diff
@@ -0,0 +1,276 @@
+|--------------------------------------------|-----------|------------|------------|---------|-----------|-------|
+| `Peers() IPeerSet`                         | x         | x          |            |         |           | x     |
+| `NumPeers() (int, int, int)`               |           | x          |            |         |           | x     |
+| `Broadcast(Envelope) chan bool`            | x         | x          | x          |         |           |       |
+| `MarkPeerAsGood(Peer)`                     | x         |            |            |         |           |       |
+| `StopPeerForError(Peer, interface{})`      | x         | x          | x          | x       | x         | x     |
+| `StopPeerGracefully(Peer)`                 |           |            |            |         |           | x     |
+| `Reactor(string) Reactor`                  |           | x          |            |         |           |       |
+
+The above list is not exhaustive as it does not include all the `Switch` methods
+invoked by the PEX reactor, a special component that should be considered part
+of the p2p layer. This document does not cover the operation of the PEX reactor
+as a connection manager.
+
+### Peers State
+
+The first two methods in the switch API allow reactors to query the state of
+the p2p layer: the set of connected peers.
+
+    func (sw *Switch) Peers() IPeerSet
+
+The `Peers()` method returns the current set of connected peers.
+The returned `IPeerSet` is an immutable concurrency-safe copy of this set.
+Observe that the `Peer` handlers returned by this method were previously
+[added to the reactor][reactor-addpeer] via the `InitPeer(Peer)` method,
+but not yet removed via the `RemovePeer(Peer)` method.
+Thus, a priori, reactors should already have this information.
+
+    func (sw *Switch) NumPeers() (outbound, inbound, dialing int)
+
+The `NumPeers()` method returns the current number of connected peers,
+distinguished between `outbound` and `inbound` peers.
+An `outbound` peer is a peer the node has dialed to, while an `inbound` peer is
+a peer the node has accepted a connection from.
+The third field `dialing` reports the number of peers to which the node is
+currently attempting to connect, so not (yet) connected peers.
+
+### Broadcast
+
+The switch provides, mostly for historical or retro-compatibility reasons,
+a method for sending a message to all connected peers:
+
+    func (sw *Switch) Broadcast(e Envelope) chan bool
+
+The `Broadcast()` method is not blocking and returns a channel of booleans.
+For every connected `Peer`, it starts a background thread for sending the
+message to that peer, using the `Peer.Send()` method
+(which is blocking, as detailed in [Send Methods](#send-methods)).
+The result of each unicast send operation (success or failure) is added to the
+returned channel, which is closed when all operations are completed.
```

Comment:
is there a way to connect the result with the peer?

Reply:
No. Essentially, the only point is to know how many sends have failed.

Reply:
Do reactors use this info?

Reply:
No, all the calls in Consensus, State Sync, and Block Sync reactors ignore the return value.

Reply:
So, candidate for removal :-)

Reply:
So, candidate for removal :-)
(not in this PR)

Reply:
Highlighted this information (no one is using this returning value) in the document.